File: vjudge/closest_pair.py
from math import sqrt

def distancia_sh(pontos, n):
    pontos.sort(key=lambda x:x[0])
    return sqrt(distancia_rec_sh(pontos, 0, n-1))

def distancia_rec_sh(pontos, p, r):
    if r <= p + 2:
        return forca_bruta(pontos, p, r)
    q = (p + r)//2
    de = distancia_rec_sh(pontos, p, q)
    dd = distancia_rec_sh(pontos, q + 1, r)
    intercale(pontos, p, q, r)
    return combine(pontos, pontos[q][0], p, r, de, dd)

def forca_bruta(pontos, p, r):
    dist2 = 100000000
    brute = pontos[p:r+1]
    brute.sort(key=lambda x:x[1])

    for i in range(p, r+1):
        pontos[i] = brute[i-p]

    for i in range(p, r+1):
        for j in range(i+1, r+1):
            aux = distancia2(pontos[i], pontos[j])
            if aux < dist2:
                dist2 = aux
    return dist2

def distancia2(a, b):
    return (b[0] - a[0])**2 + (b[1] - a[1])**2

def intercale(pontos, p, q, r):
    i = p 
    j = q + 1
    intercalado = []

    while i <= q and j <= r:
        while i <= q and pontos[i][1] <= pontos[j][1]:
            intercalado.append(pontos[i])
            i += 1

        while j <= r and pontos[j][1] <= pontos[i][1]:
            intercalado.append(pontos[j])
            j += 1
    
    while i <= q:
        intercalado.append(pontos[i])
        i += 1

    while j <= r:
        intercalado.append(pontos[j])
        j += 1
    
    for i in range(p, r+1):
        pontos[i] = intercalado[i-p]

def combine(pontos, x, p, r, de, dd):
    d = min(de,dd)
    (f,t) = candidatos(pontos, x, p, r, d)
    for i in range(t):
        for j in range(i+1, min(i+7, t)):
            dlinha = distancia2(f[i], f[j])
            if dlinha < d:
                d = dlinha
    return d

def candidatos(pontos, x, p, r, d):
    t = 0
    f = []
    for k in range(p, r+1):
        if abs(x - pontos[k][0]) * abs(x - pontos[k][0]) < d:
            f.append(pontos[k])
            t += 1
    return f, t

def main():
    n = int(input())
    while n != 0:
        pontos = []
        for i in range(n):
            coord = input().split()
            x = int(coord[0])
            y = int(coord[1])
            pontos.append((x,y))
        d = distancia_sh(pontos, n)
        if d >= 10000:
            print("INFINITY")
        else:    
            print(format(d, '.4f'))
        n = int(input())

main()

File: vjudge/ClosestPair.java
import java.util.Scanner; 
import java.lang.Math;
import java.util.Arrays;
import java.util.Comparator;

public class ClosestPair {

    public double dist;

    public ClosestPair(Ponto[] pontos, int n) {
        Arrays.sort(pontos, Ponto.X_ORDER);
        dist = Math.sqrt(distanciaRecSH(pontos, 0, n-1));
    }

    private double distanciaRecSH(Ponto[] pontos, int p, int r) {
        if (r <= p + 2) {
            return forcaBruta(pontos, p, r);
        }
        int q = (int) Math.floor((p + r)/2);
        double de = distanciaRecSH(pontos, p, q);
        double dd = distanciaRecSH(pontos, q + 1, r);
        intercale(pontos, p, q, r);
        return combine(pontos, pontos[q].x, p, r, de, dd);
    }

    private double forcaBruta(Ponto[] pontos, int p, int r) {
        double dist2 = 100000000;
        Ponto[] brute = slice(pontos, p, r+1);
        Arrays.sort(brute, Ponto.Y_ORDER);

        for (int i = p; i < r+1; i++) {
            pontos[i] = brute[i-p];
        }

        for (int i = p; i < r+1; i++) {
            for (int j = i+1; j < r+1; j++) {
                double aux = pontos[i].distancia2(pontos[j]);
                if (aux < dist2) {
                    dist2 = aux;
                }
            }
        }
        return dist2;
        
    }

    private static Ponto[] slice(Ponto[] arr, int start, int end) { 
        Ponto[] slice = new Ponto[end - start]; 
        for (int i = 0; i < slice.length; i++) { 
            slice[i] = arr[start + i]; 
        } 
        return slice; 
    } 

    private void intercale(Ponto[] pontos, int p, int q, int r) {
        int i = p; 
        int j = q + 1;
        Ponto[] intercalado = new Ponto[r - p + 1];
        int k = 0;

        while (i <= q && j <= r) {
            while (i <= q && pontos[i].y <= pontos[j].y) {
                intercalado[k++] = pontos[i++];
            }
        
            while (j <= r && pontos[j].y <= pontos[i].y) {
                intercalado[k++] = pontos[j++];
            }
        }

        while (i <= q) {
            intercalado[k++] = pontos[i++];
        }
        while (j <= r) {
            intercalado[k++] = pontos[j++];
        }

        for (i = p; i < r+1; i++) {
            pontos[i] = intercalado[i-p];
        }
    }

    private double combine(Ponto[] pontos, double x, int p, int r, double de, double dd) {
        double d = Math.min(de, dd);
        int[] ref = new int[1];
        Ponto[] f = new Ponto[r+1 - p];
        f = candidatos(pontos, x, p, r, d, ref);
        int t = ref[0];
        for (int i = 0; i < t; i++) {
            for (int j = i+1; j < Math.min(i+7, t); j++) {
                double dlinha = f[i].distancia2(f[j]);
                if (dlinha < d) {
                    d = dlinha;
                }
            }
        }
        return d;
    }

    private Ponto[] candidatos(Ponto[] pontos, double x, int p, int r, double d, int[] ref) {
        int t = 0;
        Ponto[] f = new Ponto[r+1 - p];
        for (int k = p; k < r+1; k++) {
            if (Math.abs(x - pontos[k].x) * Math.abs(x - pontos[k].x) < d) {
                f[t] = pontos[k];
                t++;
            }    
        }
        ref[0] = t;
        
        return f;
    }

    public static void main(String[] args) {
        Scanner s = new Scanner(System.in);
        int n = Integer.parseInt(s.nextLine());
        while (n != 0) {
            Ponto pontos[] = new Ponto[n];
            for (int i = 0; i < n; i++) {
                String[] coord = s.nextLine().split(" "); 
                Ponto ponto = new Ponto(Double.parseDouble(coord[0]), Double.parseDouble(coord[1]));
                pontos[i] = ponto;
            }
            double d = (new ClosestPair(pontos, n)).dist;
            if (d >= 10000) {
                System.out.println("INFINITY");
            } else {
                System.out.println(d);
            }
            n = Integer.parseInt(s.nextLine());
        }
    }
}

class Ponto implements Comparable<Ponto> {

    public double x;
    public double y;
    public static final Comparator<Ponto> X_ORDER = new XOrder();
    public static final Comparator<Ponto> Y_ORDER = new YOrder();

    public Ponto(double x, double y) {
        this.x = x;
        this.y = y;
    }

    public double distancia2(Ponto other) {
        return Math.pow(this.x - other.x, 2) + Math.pow(this.y - other.y, 2);
    }

    public int compareTo(Ponto that) {
        if (this.x < that.x) return -1;
        if (this.x > that.x) return +1;
        if (this.y < that.y) return -1;
        if (this.y > that.y) return +1;
        return 0;
    }

    private static class YOrder implements Comparator<Ponto> {
        public int compare(Ponto p, Ponto q) {
            if (p.y < q.y) return -1;
            if (p.y > q.y) return +1;
            return 0;
        }
    }

    private static class XOrder implements Comparator<Ponto> {
        public int compare(Ponto p, Ponto q) {
            return p.compareTo(q);
        }
    }
}

class WrapInt {
    public int value;
}